src/bucket.ts

```diff
@@ -3174,6 +3185,201 @@ class Bucket extends ServiceObject {
+   * @param {?Error} err Request error, if any.
+   * @param {array} resp The list of object responses for every uploaded file.
+   */
+
+  /**
+   * Upload a directory to the bucket. This is a convenience method for multiple
+   * {@link Bucket#upload}.
+   *
+   * You can specify whether or not an upload is resumable by setting
+   * `options.resumable`. *Resumable uploads are enabled by default if your
+   * input file is larger than 5 MB.*
+   *
+   * For faster crc32c computation, you must manually install
+   * [`fast-crc32c`](http://www.gitnpm.com/fast-crc32c):
+   *
+   *     $ npm install --save fast-crc32c
+   *
+   * @see [Upload Options (Simple or Resumable)]{@link https://cloud.google.com/storage/docs/json_api/v1/how-tos/upload#uploads}
+   * @see [Objects: insert API Documentation]{@link https://cloud.google.com/storage/docs/json_api/v1/objects/insert}
+   *
+   * @param {string} directoryPathString The fully qualified path to the
+   *     directory you wish to upload to your bucket.
+   * @param {UploadDirectoryOptions} [options] Configuration options.
+   * @param {UploadDirectoryCallback} [callback] Callback function.
+   * @returns {Promise<UploadDirectoryResponse>}
+   *
+   * @example
+   * const {Storage} = require('@google-cloud/storage');
+   * const storage = new Storage();
+   * const bucket = storage.bucket('albums');
+   *
+   * //-
+   * // Upload a directory from a local path.
+   * //-
+   * bucket.uploadDirectory('/local/path/myfolder', function(err, resp) {
+   *   // Your bucket now contains:
+   *   // - "myfolder/" directory (with the contents of `/local/path/myfolder/')
+   *
+   *   // `file` is an instance of a File object that refers to your new file.
+   * });
+   *
+   *
+   * //-
+   * // You can also have files gzip'd on the fly.
+   * //-
+   * bucket.uploadDirectory(path.join(__dirname, 'index', { gzip: true }, function(err, resp) {
+   *   // Your bucket now contains:
+   *   // - "index/" directory (automatically compressed with gzip)
+   *
+   *   // Downloading the file with `file.download` will automatically decode
+   * the
+   *   // file.
+   * });
+   *
+   *
+   * //-
+   * // If the callback is omitted, we'll return a Promise.
+   * //-
+   * bucket.uploadDirectory('local-directory-path').then(function(data) {
+   *   const resp = data[0];
+   * });
+   */
+
+  uploadDirectory(
+    directoryPathString: string,
+    optionsOrCallback?: UploadDirectoryOptions | UploadDirectoryCallback,
+    callback?: UploadDirectoryCallback
+  ): Promise<UploadDirectoryResponse> | void {
+    // tslint:disable-next-line:no-any
+    if ((global as any)['GCLOUD_SANDBOX_ENV']) {
+      return;
+    }
+
+    const options =
+      typeof optionsOrCallback === 'object' ? optionsOrCallback : {};
+    callback =
+      typeof optionsOrCallback === 'function' ? optionsOrCallback : callback;
+
+    let stat: fs.Stats | Error;
+    try {
+      stat = fs.lstatSync(directoryPathString);
```

Comment:
I think we might want to use the async version of this? @bcoe @stephenplusplus thoughts?